torch/csrc/jit/codegen/cuda/arith.cpp

```diff
@@ -242,28 +227,81 @@ TensorView* arithOpOverloads(
              vals[3]->template as<Val>())
       ->template as<TensorView>();
 }
+
+// Type promotion logic for binary operators
+DataType getOutputType(BinaryOpType op_type, Val* v1, Val* v2) {
+  DataType v1_dtype = v1->getDataType().value();
+  DataType v2_dtype = v2->getDataType().value();
+
+  // If we have a tensor view in one argument but a scalar in the other, don't
+  // type promote, just use the tensorview type
+  if (v1->isA<TensorView>() && !v2->isA<TensorView>()) {
+    v2_dtype = v1_dtype;
+  }
+  if (v2->isA<TensorView>() && !v1->isA<TensorView>()) {
+    v1_dtype = v2_dtype;
+  }
+
+  const bool floating_input =
+      isFloatingPointType(v1_dtype) || isFloatingPointType(v2_dtype);
+
+  const bool integer_input =
+      isIntegralType(v1_dtype) || isIntegralType(v2_dtype);
+
+  const bool all_integer_input =
+      isIntegralType(v1_dtype) && isIntegralType(v2_dtype);
+
+  if (isIntegerOp(op_type) || (alsoBooleanOperator(op_type) && integer_input)) {
+    // If integer op or maybe bool op with integer inputs meaning binary op
+    if (integer_input && all_integer_input) {
+      return promote_type(v1_dtype, v2_dtype);
+    } else if (integer_input && !all_integer_input) {
+      return isIntegralType(v1_dtype) ? v1_dtype : v2_dtype;
+    } else {
+      return DataType::Int;
```

Comment:
At this point we have integer operation with inputs that are not integer. Maybe assert false here?